site/src/pages/release-notes/1.17.0.mdx

```diff
@@ -0,0 +1,309 @@
+  ByteStreamMessage pathContent =
+      StreamMessage.builder(path)
+                   .bufferSize(1024)
+                   .alloc(alloc)
+                   .executor(executor)
+                   .build();
+  // Slice a range of a file data
+  ByteStreamMessage partialContent = pathContent.range(100, 200);
+  ```
+- You can now map a specific exception to a `LogLevel` in <type://LoggingService> and
+  <type://LoggingClient>. #3400 #4090
+  ```java
+  LoggingService.builder()
+                .responseLogLevel(IllegalStateException.class,
+                                  LogLevel.WARN)
+                .responseLogLevel(IllegalArgumentException.class,
+                                  LogLevel.ERROR)
+                ...
+  ```
+- You can now map a specific exception to an <type://HttpResponse>. #4279 #4283
+  ```java
+  HttpResponse recovered =
+      response.recover(IllegalStateException.class,
+                       cause -> HttpResponse.of("Fallback"));
+  ```
+- You can now enable automatic compression for a gRPC response. #4258 #4266
+  ```java
+  GrpcService.builder()
+             .addService(new MyGrpcService())
+             .autoCompression(true)
+             .build();
+  ```
+  - This only works when the client sends a `grpc-accept-encoding` header and
+    the `CompressorRegistry` has the compressor.
+- You can now add a hook to <type://Server> and <type://ClientFactory> that is called when the
+  JVM shuts down. #4015 #4043
+  ```java
+  server.closeOnJvmShutdown(() -> {
+    System.err.println("Server is stopping soon.");
+  }).thenRun(() -> {
+    System.err.println("Server has stopped.");
+  });
+  ```
+- You can now customize internal Netty `Pipeline` using
+  <type://ClientFactoryOptions#CHANNEL_PIPELINE_CUSTOMIZER>. #3907 #4260
+  - This is an advanced feature and not recommended to use if you are not familiar with
+    Armeria and Netty internals.
+
+### Kotlin
+- You can now use the Kotlin nullable type (`?`) to indicate a nullable parameter in
+  an annotated service. #4144 #4225
+  ```kotlin
+  serverBuilder.apply {
+    annotatedService(object {
+      // Both `/foo` and `/foo?a=bar` are allowed.
+      @Get("/foo")
+      fun foo(@Param a: String?) = // 👈👈👈
+        HttpResponse.of("a: $a")
+    })
+  }
+  ```
+
+## 📈 Improvements
+
+- The <type://EndpointSelectionTimeoutException> is thrown if timeout occurred when selecting
+  an <type://Endpoint> from an <type://EndpointGroup>. #4269
+- You can now set multiple <typeplural://AccessLogWriter> to <type://ServerBuilder>
+  and <type://ServiceBindingBuilder>. #4280
+- You can now give a performance optimization hint to a client by setting an <type://ExchangeType> to a
+  <type://RequestOptions>.
+  The <type://ExchangeType> is configured automatically for:
+  - A gRPC client
+  - A Thrift client
+  - A <type://WebClient> when it's used with <type://ResponseAs>. #4236
+- Better performance for unary gRPC call by introducing `UnaryServerCall`. #4192
```

Comment:
`UnaryServerCall` is implementation detail that could be omitted.

Reply:
Removed. 😉